### PR TITLE
DataExporter: Enhance DataExporter to support other components #4414

### PR DESCRIPTION
**Pull request by ravihariharan**
Creating a PR for discussion for support for DataExporter for use with other components

`Exporter<TABLE extends UIComponent>`
i could only extends it to UIComponent  since UIData would have excluded all UITree components. 

Open to review and suggestion to make this PR better.

### CODE REVIEW

**Comment by tandraschko** [comment]
lets discuss this after the release. RC1 is coming soon.

**Comment by ravihariharan** [comment]
@tandraschko  i was trying to sneak it to 6.3  :). It was quick refactor for rename a class and extract a interface. 

I guess we can wait if RC1 is round the corner.

**Comment by ravihariharan** [comment]
@tandraschko let me know you thought of how this could be extended.

**Comment by tandraschko** [comment]
will do it when i have more time, probably in ~2-3 weeks

**Comment by melloware** [comment]
@ravihariharan Would this PR fix this issue by allowing export of TreeTable?  https://github.com/primefaces/primefaces/issues/10

If so this would be a big win.

**Comment by ravihariharan** [comment]
@melloware  the refactor will allow any developer to provide a custom export implementation. We should be able to use to use it for TreeTable .  

DataTable will have default export support for PDF, CVS, XML. At a later date we can also build the default export support for TreeTable.

File: src/main/java/org/primefaces/component/export/Exporter.java [hunk 1]
```diff
@@ -16,320 +16,26 @@
...
-import javax.faces.component.html.HtmlGraphicImage;
+import javax.faces.component.UIComponent;
 import javax.faces.context.FacesContext;
-import javax.faces.convert.Converter;
 
-import org.primefaces.component.celleditor.CellEditor;
-import org.primefaces.component.datatable.DataTable;
-import org.primefaces.component.overlaypanel.OverlayPanel;
-import org.primefaces.util.ComponentUtils;
+public interface Exporter<TABLE extends UIComponent> {
```
Comment (Rapster) [inline src/main/java/org/primefaces/component/export/Exporter.java:25]: Rename into `T`, as you mentionned it does not necessarily need to be a `DataTable` or a `TABLE` at all

File: src/main/java/org/primefaces/model/charts/ChartDataSet.java [hunk 2]
```diff
@@ -18,11 +18,35 @@
 import java.io.IOException;
 import java.io.Serializable;
 
+/**
```
Comment (Rapster) [inline src/main/java/org/primefaces/model/charts/ChartDataSet.java:21]: Why chartjs files have been modified in this PR? Wrong pull?

**Comment by Rapster** [comment]
I like this PR, though see my comments please

**Comment by ravihariharan** [comment]
@Rapster  i could not find any comments in the commit.

https://github.com/primefaces/primefaces/pull/4416/commits/7c6828f50c5d4115c006a791d365075fdabf5b79

**Comment by Rapster** [comment]
Don't look in the commit but PR itself

**Comment by ravihariharan** [comment]
@Rapster  the only comments  i see 
`I like this PR, though see my comments please`
`Don't look in the commit but PR itself`

checked both the issue and PR :(
https://github.com/primefaces/primefaces/issues/4414
https://github.com/primefaces/primefaces/pull/4416

**Review by Rapster — CHANGES_REQUESTED** [review]

File: src/main/java/org/primefaces/model/charts/ChartDataSet.java (hunk 2 above)
Comment (ravihariharan) [reply src/main/java/org/primefaces/model/charts/ChartDataSet.java:21]: i had merged @melloware  code to test and incorrectly merged it to my PR .
@melloware code is already merged to master.

File: src/main/java/org/primefaces/component/export/Exporter.java (hunk 1 above)
Comment (ravihariharan) [reply src/main/java/org/primefaces/component/export/Exporter.java:25]: i messed up my github repository. can this be fixed with a new PR after this one is merged ?  or i will redo with a new repository and a new PR.

Comment (Rapster) [reply src/main/java/org/primefaces/component/export/Exporter.java:25]: There are plenty ways to about it, but you don't need to create a new repo.  A new branch should do (but I think you can even solve your problem without doing that). Play with it :wink:

Comment (ravihariharan) [reply src/main/java/org/primefaces/component/export/Exporter.java:25]: @Rapster i was able to get the orphaned  PR using   
```
git fetch https://github.com/primefaces/primefaces.git refs/pull/4416/head:4416PR
git checkout 4416PR

```
These changes are now
image

The above will create a PR with a new ID so not sure if there is a better way?

Comment (Rapster) [reply src/main/java/org/primefaces/component/export/Exporter.java:25]: It's okay, just close the old PR

**Comment by awenger63** [comment]
Hi, do you know if this could be use whith pe:sheet ?

**Comment by ravihariharan** [comment]
New PR : DataExporter: Enhance DataExporter to support other components #4414 #4949

Review comments fixed.

**Comment by ravihariharan** [comment]
> pe:sheet



> Hi, do you know if this could be use whith pe:sheet ?

The objective of this PR is to provide only a interface for developer to implement a custom export for any component. 

```
https://primefaces.github.io/primefaces/7_0/#/components/dataexporter
customExporter
```

**Pull request closed without merging**